flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/aggfunctions/SortAggFunction.scala

```diff
@@ -0,0 +1,166 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.table.functions.aggfunctions
+
+import java.math.BigDecimal
+import java.util.{List => JList}
+
+import org.apache.flink.api.common.typeinfo.{BasicTypeInfo, TypeInformation}
+import org.apache.flink.api.java.tuple.{Tuple2 => JTuple2}
+import org.apache.flink.api.java.typeutils.TupleTypeInfo
+import org.apache.flink.table.functions.{Accumulator, AggregateFunction}
+import scala.collection.mutable.ArrayBuffer
+import org.apache.flink.types.Row
+import org.apache.flink.table.runtime.aggregate.UntypedOrdering
+import org.apache.flink.table.runtime.aggregate.MultiOutputAggregateFunction
+import java.util.{ List => JList,ArrayList }
+import org.apache.flink.api.java.typeutils.RowTypeInfo
+import org.apache.flink.api.java.typeutils.ListTypeInfo
+import org.apache.calcite.rel.RelFieldCollation.Direction
+
+/** The initial accumulator for Sort aggregate function */
+class SortAccumulator extends ArrayBuffer[JTuple2[Row,Row]] with Accumulator with Serializable
+
+/**
+  * Base class for built-in Min aggregate function
+  *
+  * @tparam K the type for the key sort type
+  * @tparam T the type for the aggregation result
+  */
+abstract class SortAggFunction[K,T](
+    val keyIndexes: Array[Int],
+    val keySortDirections: Array[Direction],
+    val orderings: Array[UntypedOrdering]) extends MultiOutputAggregateFunction[T] {
+
+  override def createAccumulator(): Accumulator = {
+    val acc = new SortAccumulator
+      
+    acc
+  }
+
+  override def accumulate(accumulator: Accumulator, value: Any): Unit = {
+    if (value != null) {
+      val v = value.asInstanceOf[Row]
+      val acc = accumulator.asInstanceOf[SortAccumulator]
+      
+      var i = 0
+      //create the (compose) key of the new value
+      val keyV = new Row(keyIndexes.size)
+      while (i<keyIndexes.size) {
+        keyV.setField(i, v.getField(keyIndexes(i)))
+        i += 1
+      }
+        
+      var j = 0
+      while (j<acc.size) {
+        i = 0
+        while (i<keyIndexes.size) {
+          
+          val compareResult = if(keySortDirections(i) == Direction.ASCENDING) {
+            orderings(i).compare(acc(j).f0.getField(i),keyV.getField(i))
+          } else {
+            orderings(i).compare(keyV.getField(i),acc(j).f0.getField(i))
+          }
+          
+          compareResult match {
+          case 0 => i += 1 //same key and need to sort on consequent keys 
+          case g if g > 0 => {
+            acc.insert(j, new JTuple2(keyV,v)) //add new element in place
```

Comment:
Insertion in an `ArrayBuffer` should be expensive, because all following entries need to be moved.